{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Objective\n",
    "We got a uBAM file generated by sequencing of a PCR sample. In the process of the analysis the uBAM got converted to SAM-format.\n",
    "The file contains all sorts of reads with various lengths, presence or not of the PCR primers (complete and incomplete sequences), forward and reverse reads, and chunk reads.\n",
    "We want to generate a fasta file that contains only the complete region of interest in a certain range of lenghts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import all the needed packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import string\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio.Seq import reverse_complement as RC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create a sam file for the purpose of this exercise\n",
    "The mock file matches the format and content of a real sam file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define functions to create random genomic sequences, random names and the content of the sam file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_random_seq(read_length=10, cg_cont=.5):\n",
    "    \"\"\"This function creates a random DNA sequence of custom length and of custom GC content.\n",
    "    \n",
    "    Arguments:\n",
    "    read_length: length of the sequence; default = 10 bases.\n",
    "    cg_cont: fraction of G and C nucleotides; default = 0.5.\n",
    "    \"\"\"\n",
    "    \n",
    "    base_p = [(1 - cg_cont) / 2, (1 - cg_cont) / 2, cg_cont / 2, cg_cont / 2] # List of probability for each nucleotide\n",
    "    return ''.join(np.random.choice(['A', 'T', 'C', 'G'], p=base_p, size=read_length)) # Create a random sequence (taking into account the desired proportion of each nucleotide) of defined length\n",
    "\n",
    "\n",
    "def get_random_name(size=10, chars=string.ascii_uppercase + string.digits):\n",
    "    \"\"\"This function creates a random name of custom length and composed of upper case letters and digits by default.\n",
    "    \n",
    "    Arguments:\n",
    "    size: length of the name; default = 10 characters.\n",
    "    chars: types of characters included in the name; default = upper case letters and digits.\n",
    "    \"\"\"\n",
    "    \n",
    "    return ''.join(random.choice(chars) for _ in range(size)) # Create a random name of defined composition and length\n",
    "\n",
    "\n",
    "def create_sam(n_reads=100, seq1='', seq2='', min_roi=200, max_roi=500):\n",
    "    \"\"\"This function creates a custom number of random sequencing reads in sam format. Reads incorporate two possible custom sequences.\n",
    "    Only names and sequences are actually created. All other sam parameters (such as flag, PHRED score...) are left undefined using the '-' character.\n",
    "    \n",
    "    Arguments:\n",
    "    n_reads: number of reads to create; default = 100.\n",
    "    seq1: custom sequence 1 to incorporate in the reads; default = empty string.\n",
    "    seq2: custom sequence 2 to incorporate in the reads; default = empty string.\n",
    "    min_roi: min length of the region of intestest\n",
    "    max_roi: max length of the region of intestest\n",
    "    \"\"\"\n",
    "\n",
    "    reads = []\n",
    "    names = []\n",
    "\n",
    "    for _ in range(1, n_reads+1): ## range(n_reads)\n",
    "        reads.append(''.join([get_random_seq(random.randint(1, 50)),            # Append the list of reads composed of a random sequence of maximum 50 nucleotides, then\n",
    "                              seq1,                                             # the defined sequence 1, then\n",
    "                              get_random_seq(random.randint(min_roi, max_roi)), # a random sequence of min_roi to max_roi nucleotides, then\n",
    "                              seq2,                                             # the defined sequence 2, then\n",
    "                              get_random_seq(random.randint(1, 50))]))          # a random sequence of maximum 50 nucleotides\n",
    "        names.append(get_random_name()) # Use the get_random_name() function to append the list of names\n",
    "\n",
    "    d = {'QNAME': names,          # Create a dictionary composed of all fields defining a sam file and their content\n",
    "         'FLAG': ['-'] * n_reads,\n",
    "         'RNAME': ['-'] * n_reads,\n",
    "         'POS': ['-'] * n_reads,\n",
    "         'MAPQ': ['-'] * n_reads,\n",
    "         'CIGAR': ['-'] * n_reads,\n",
    "         'RNEXT': ['-'] * n_reads,\n",
    "         'PNEXT': ['-'] * n_reads,\n",
    "         'TLEN': ['-'] * n_reads,\n",
    "         'SEQ': reads,\n",
    "         'QUAL': ['-'] * n_reads}\n",
    "    \n",
    "    return pd.DataFrame(d) # Make the dictionary a dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create the sam file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 2500 entries, 0 to 999\n",
      "Data columns (total 11 columns):\n",
      "QNAME    2500 non-null object\n",
      "FLAG     2500 non-null object\n",
      "RNAME    2500 non-null object\n",
      "POS      2500 non-null object\n",
      "MAPQ     2500 non-null object\n",
      "CIGAR    2500 non-null object\n",
      "RNEXT    2500 non-null object\n",
      "PNEXT    2500 non-null object\n",
      "TLEN     2500 non-null object\n",
      "SEQ      2500 non-null object\n",
      "QUAL     2500 non-null object\n",
      "dtypes: object(11)\n",
      "memory usage: 234.4+ KB\n",
      "GGTGGCGACCGTGTAGTCCGCGGATTAATGGTATAaaaatAACACACGTTAGAACCGTATCGAGCATACAGCGGTGTTTCACCGGTGTTCACACCTCGCACCTGCCAAGACTACGGAGCCCCGTCGATCATAAACAGTGAGGCTGAGGCCGAGTGGCACCACGGTTGCGAGCCGGGCAGCCCGATCACATCGTTCCTTAAACACATAGGTACAATATCTCGCGTTACTGGTAAATATGTGTCAAATTAAGATCCTTAAGAGGCTGAGGCGTTTACTACTACGCGTTAAGGAATGTATGTGAACGAGACTGAATGATTTTCACTTTCGTAATCATGATAATACGTTGGCTGTGTCATCTGCGCTGGAGAATGTAGCAGCTTTGCTTTTTTGTGAGCAGAAGCTTCCACGAGGTGATTGGGAATGGCCTATGAATTATTGGCACCAAAGGGAGGCAGGCGACTGAAACATTACCCGGAGTTATGGGTCACGAcggggCCGAAGACGGTTTGGATTCCTCGCACCGCCGTG\n"
     ]
    }
   ],
   "source": [
    "forward = 'aaaat' # Sequence of the forward primer of the region of interest\n",
    "reverse = 'ccccg' # Sequence of the reverse primer of the region of interest\n",
    "\n",
    "d = create_sam(n_reads=500, seq1=forward, seq2=RC(reverse))  # Complete reads\n",
    "d = pd.concat([d, create_sam(n_reads=500, seq1=reverse, seq2=RC(forward))])  # Complete reverse complement reads\n",
    "d = pd.concat([d, create_sam(n_reads=250, seq1=forward)])  # Incomplete reads containing forward primer only\n",
    "d = pd.concat([d, create_sam(n_reads=250, seq1=reverse)])  # Incomplete reads containing reverse primer only\n",
    "d = pd.concat([d, create_sam(n_reads=1000, max_roi=3000)])  # Completely random reads\n",
    "\n",
    "d.info() # Get a summary of the properties of the dataframe\n",
    "d.head() # Visualize head of the dataframe\n",
    "print(d.iloc[0,9]) # Print complete sequence of first read\n",
    "\n",
    "testfile = \"./test.sam\" # Set sam file name and location\n",
    "d.to_csv(testfile, sep='\\t', index=False, header=False) # Write file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perform the analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define a function that trims sequencing reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def shorten_read(read, primer, n=500):\n",
    "    \"\"\"This function trims reads before the starting position of a primer sequence and after a custom length.\n",
    "    \n",
    "    Arguments:\n",
    "    read: a genomic sequence.\n",
    "    primer: the sequence of a primer to be found in the read (defines the starting postition of the retained sequence).\n",
    "    n: final length of the read; default = 500 bases.\n",
    "    \"\"\"\n",
    "    read = read[read.find(primer):read.find(primer) + n] # Slice read from the position where primer sequence is found to that position + n bases\n",
    "    return read # Return trimmed read"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "infile = \"test.sam\" # Set input file\n",
    "outfile = \"test.clean.fa\" # Set output file\n",
    "\n",
    "# If analysing a real file (not created previoulsy in this code), the following 2 commands have to be run\n",
    "#forward = \"\" # Write forward primer sequence\n",
    "#reverse = \"\" # Write reverse primer sequence\n",
    "\n",
    "max_len = 400 # Set maximum length of the region of interest\n",
    "min_len = 300 # Set minimum length of the region of interest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DTGYTMGRKV</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>GGTGGCGACCGTGTAGTCCGCGGATTAATGGTATAaaaatAACACA...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WHT06NUKLX</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CCCTGCGCCATATAACCGTCCTCaaaatGGAGAACCGGCGGTCCGT...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>841NA94S90</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ATATCGATTAACACGTCTAGCTAGAGTCAGGAATTTaaaatTGAGC...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FJSAGCII11</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>GAACTGAGGATTCAaaaatACAGCGCTCGGGACTCGTTCTCGTGTG...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FM7E568SM9</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>CCGTCCTATCTAGCGTCCCTTAGTGTCAGTCGaaaatGTTACTTAA...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           0  1  2  3  4  5  6  7  8   \\\n",
       "0  DTGYTMGRKV  -  -  -  -  -  -  -  -   \n",
       "1  WHT06NUKLX  -  -  -  -  -  -  -  -   \n",
       "2  841NA94S90  -  -  -  -  -  -  -  -   \n",
       "3  FJSAGCII11  -  -  -  -  -  -  -  -   \n",
       "4  FM7E568SM9  -  -  -  -  -  -  -  -   \n",
       "\n",
       "                                                  9  10  \n",
       "0  GGTGGCGACCGTGTAGTCCGCGGATTAATGGTATAaaaatAACACA...  -  \n",
       "1  CCCTGCGCCATATAACCGTCCTCaaaatGGAGAACCGGCGGTCCGT...  -  \n",
       "2  ATATCGATTAACACGTCTAGCTAGAGTCAGGAATTTaaaatTGAGC...  -  \n",
       "3  GAACTGAGGATTCAaaaatACAGCGCTCGGGACTCGTTCTCGTGTG...  -  \n",
       "4  CCGTCCTATCTAGCGTCCCTTAGTGTCAGTCGaaaatGTTACTTAA...  -  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_table(infile, sep=\"\\t\", header=None) # Read input file\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select columns of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 2500 entries, 0 to 2499\n",
      "Data columns (total 2 columns):\n",
      "name    2500 non-null object\n",
      "seq     2500 non-null object\n",
      "dtypes: object(2)\n",
      "memory usage: 39.1+ KB\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DTGYTMGRKV</td>\n",
       "      <td>GGTGGCGACCGTGTAGTCCGCGGATTAATGGTATAaaaatAACACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WHT06NUKLX</td>\n",
       "      <td>CCCTGCGCCATATAACCGTCCTCaaaatGGAGAACCGGCGGTCCGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>841NA94S90</td>\n",
       "      <td>ATATCGATTAACACGTCTAGCTAGAGTCAGGAATTTaaaatTGAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FJSAGCII11</td>\n",
       "      <td>GAACTGAGGATTCAaaaatACAGCGCTCGGGACTCGTTCTCGTGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FM7E568SM9</td>\n",
       "      <td>CCGTCCTATCTAGCGTCCCTTAGTGTCAGTCGaaaatGTTACTTAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         name                                                seq\n",
       "0  DTGYTMGRKV  GGTGGCGACCGTGTAGTCCGCGGATTAATGGTATAaaaatAACACA...\n",
       "1  WHT06NUKLX  CCCTGCGCCATATAACCGTCCTCaaaatGGAGAACCGGCGGTCCGT...\n",
       "2  841NA94S90  ATATCGATTAACACGTCTAGCTAGAGTCAGGAATTTaaaatTGAGC...\n",
       "3  FJSAGCII11  GAACTGAGGATTCAaaaatACAGCGCTCGGGACTCGTTCTCGTGTG...\n",
       "4  FM7E568SM9  CCGTCCTATCTAGCGTCCCTTAGTGTCAGTCGaaaatGTTACTTAA..."
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = data[[0,9]] # Select 1st and 10th columns of the dataframe\n",
    "df.columns = [\"name\",\"seq\"] # Name the columns\n",
    "df.info()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot distribution of read length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEKCAYAAAAfGVI8AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAE8BJREFUeJzt3X+wbWV93/H3xwsILSogR3sD2gMNjTCZFPXKYMhkLCga\nSYpaTMnY5CbSMmn8WavNTZ22MtOZYidVm9bREjXeOI5i0BQUG0IRdDJjkHsVEErMvVJsqXe41yr+\nmDCkkG//WOvI9vT82OfcvfY++zzv18yZvfaz117r++x17vnctZ69n52qQpLUrifNugBJ0mwZBJLU\nOINAkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGHTPrAsZx6qmn1uLi4qzLkKS5sn///m9V\n1cJ6681FECwuLrJv375ZlyFJcyXJN8ZZz0tDktQ4g0CSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ1\nziCQpMYZBJLUOINAkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQJIaZxBIUuMM\nAklqnEEgSY0zCCSpcQaBJDXOIJCkxhkEktQ4g0CSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ1ziCQ\npMYNHgRJdiT5SpLP9PfPSHJ7kgNJrk1y3NA1SJJWN40zgjcB943cfyfw7qo6C/gOcMUUapAkrWLQ\nIEhyOnAJ8IH+foALgev6VfYCrxiyBknS2oY+I3gP8M+Bv+rvPx14uKoe6+8/CJw2cA2SpDUMFgRJ\nfh44XFX7R5tXWLVWef6VSfYl2XfkyJFBahzX4p4bZ7p/SRrSkGcEFwB/L8kDwMfpLgm9BzgpyTH9\nOqcD31zpyVV1TVXtqqpdCwsLA5YpSW0bLAiq6req6vSqWgQuBz5XVa8BbgUu61fbDVw/VA2SpPXN\n4nMEvwm8JclBujGDD86gBklSbypBUFW3VdXP98v3V9V5VfXjVfXqqnp0GjVMi+MJkuaNnyyWpMYZ\nBJLUOINAkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNc4gmCA/TCZpHhkEktQ4g0CSGmcQSFLjDAJJ\napxBMAAHjSXNE4NAkhpnEEhS4wwCSWpc80Hg9XxJrWs+CCSpdQaBJDXOIJCkxhkEktQ4g6C3uOfG\nNQeOHVSWtF0ZBJLUOINAkhpnEEhS4wyCCXD8QNI8MwgkqXEGgSQ1ziCQpMYZBJLUOINgFQ4AS2qF\nQSBJjTMIJKlxBoEkNc4gWIOT0ElqwWBBkOT4JF9KcleSe5Nc1befkeT2JAeSXJvkuKFqkCStb8gz\ngkeBC6vq7wDnAi9Lcj7wTuDdVXUW8B3gigFrkCStY7AgqM4P+rvH9j8FXAhc17fvBV4xVA2SpPUN\nOkaQZEeSO4HDwM3A14GHq+qxfpUHgdOGrEGStLZBg6CqHq+qc4HTgfOAs1dabaXnJrkyyb4k+44c\nOTJkmT9iiEFgB5YlbWVTeddQVT0M3AacD5yU5Jj+odOBb67ynGuqaldV7VpYWJhGmZLUpCHfNbSQ\n5KR++QTgxcB9wK3AZf1qu4Hrh6pBkrS+Y9ZfZdN2AnuT7KALnE9U1WeS/Hfg40n+DfAV4IMD1iBJ\nWsdgQVBVdwPPXaH9frrxgi1j+TV8r+lLaomfLJakxhkEktQ4g0CSGmcQSFLjmg2CSQ8IO8AsaV41\nGwSSpM6GgyDJyUl+aohiJEnTN1YQJLktyVOTnALcBfxekncNW5okaRrGPSN4WlV9D3gV8HtV9Xy6\nKSO2vaO99u/YgaStbtwgOCbJTuAXgc8MWI8kacrGDYKrgJuAg1V1R5IzgQPDlSVJmpZx5xo6VFU/\nHCCuqvsdI5Ck7WHcM4L/OGabJGnOrHlGkOSFwE8DC0neMvLQU4EdQxY2D9YbCB59fHHPjTxw9SVD\nlyRJG7bepaHjgBP79Z4y0v49nvhyGUnSHFszCKrq88Dnk3y4qr4xpZokSVM07mDxk5NcAyyOPqeq\nLhyiKEnS9IwbBH8AvB/4APD4cOVsfV7rl7TdjBsEj1XV+watRJI0E+O+ffTTSX4jyc4kpyz9DFqZ\nJGkqxj0j2N3fvm2krYAzJ1uOJGnaxgqCqjpj6EIkSbMxVhAk+ZWV2qvq9ydbznQ4I6gkPWHcS0Mv\nGFk+HrgI+DIwl0EgSXrCuJeG3jB6P8nTgI8MUpEkaao2+53FfwGcNclCJEmzMe4Ywafp3iUE3WRz\nZwOfGKqorWppbMExBknbybhjBL89svwY8I2qenCAeiRJUzbWpaF+8rk/o5uB9GTgL4csSpI0PWMF\nQZJfBL4EvJrue4tvT+I01JK0DYx7aejtwAuq6jBAkgXgvwHXDVWYJGk6xn3X0JOWQqD3fzbw3C3L\nQV9JGv+M4I+S3AR8rL//D4DPDlOSJGma1vvO4h8HnllVb0vyKuBngABfBD46hfokSQNb7/LOe4Dv\nA1TVp6rqLVX1T+nOBt4zdHGSpOGtFwSLVXX38saq2kf3tZVaheMPkubFekFw/BqPnbDWE5M8K8mt\nSe5Lcm+SN/XtpyS5OcmB/vbkjRYtSZqc9YLgjiT/eHljkiuA/es89zHgn1XV2cD5wOuSnAPsAW6p\nqrOAW/r7kqQZWe9dQ28G/jDJa3jiD/8u4DjglWs9saoOAYf65e8nuQ84DbgUeFG/2l7gNuA3N1G7\nJGkC1gyCqnoI+Okkfxf4yb75xqr63EZ2kmQReC5wO927kJYC4lCSZ2y0aEnS5Iz7fQS3ArduZgdJ\nTgQ+Cby5qr6XZNznXQlcCfDsZz97M7tekYO4kvSjBv10cJJj6ULgo1X1qb75oSQ7+8d3AodXem5V\nXVNVu6pq18LCwpBlSlLTBguCdP/1/yBwX1W9a+ShG4Dd/fJu4PqhapAkrW/cKSY24wLgl4GvJrmz\nb/sXwNXAJ/p3Hv1PuhlNJUkzMlgQVNWf0E1HsZKLhtrvVra450YeuPqSWZchST9i7mcQlSQdHYNA\nkhpnEEhS4wwCSWqcQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQJIaZxBIUuMMAklqnEEgSY1rJgj8\nZjJJWlkzQSBJWplBIEmNMwgkqXEGgSQ1ziCQpMYZBJLUOINAkhpnEEhS4wwCSWqcQSBJjTMIJKlx\nBoEkNc4gkKTGGQSS1DiDQJIaZxBIUuMMAklqnEEgSY0zCCSpcQaBJDXOIJCkxhkEM7C458ZZlyBJ\nPzRYECT5UJLDSe4ZaTslyc1JDvS3Jw+1f0nSeIY8I/gw8LJlbXuAW6rqLOCW/r4kaYYGC4Kq+gLw\n7WXNlwJ7++W9wCuG2r8kaTzTHiN4ZlUdAuhvnzHl/UuSltmyg8VJrkyyL8m+I0eOTGSbDtJK0v9v\n2kHwUJKdAP3t4dVWrKprqmpXVe1aWFiYWoGS1JppB8ENwO5+eTdw/ZT3L0laZsi3j34M+CLwE0ke\nTHIFcDXwkiQHgJf09yVJM3TMUBuuql9a5aGLhtqnJGnjtuxgsSRpOgwCSWqcQSBJjTMIJKlxBsGU\n+aE2SVuNQSBJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQJIaZxBIUuMMAklqnEEgbQPb6YOK26kv88Ig\nkKTGGQSS1DiDQJIa10QQbMVrjkPXtNL2N7vP1Z436e2Nu85mH9vIOpMwqf3M4vd33l4jHZ0mgkCS\ntDqDQJIaZxBIUuMMAklqnEGwxSwNni2/XWmd7WK1vk6ynxvZ1hCv73oD3JMceD+abY0+d7Xlo7XW\n7/bRbnOt7U7yd2C0D6M/m93+0fwOTIJBIEmNMwgkqXEGgSQ1btsHwXa6nr7WdcSjvZ47ieu2415X\nXW/doT4UNvr6DXGdeq39Tmq95demj2a7G+370b5WK73+y9vH2e84Y0njvpZr/U4e7ZjVRv49zNq2\nDwJJ0toMAklqnEEgSY0zCCSpcQbBjG1mUGuj21hvkHatwcRxBlXXGlRb/mGbzQy4rfahnXEHTDfy\nGo/bz9VqGWdAfyPHY7PHa9z9r7W9lbY57nHc7Ies1hoMH+d12czvw3qDxhtpW2t/k/zw36QZBJLU\nOINAkhpnEEhS41JVs65hXbt27ap9+/Zt6rlb5RrcpD1w9SWb7tsDV18CbO61WW2/S+2jj2+mxqPp\n10q1wOr9XGtfa/VjnBon0Y+jdbTHYqPbXe31ntVrsdLv5NFsZ5I1jbPeJCTZX1W71ltvJmcESV6W\n5GtJDibZM4saJEmdqQdBkh3Ae4GfA84BfinJOdOuQ5LUmcUZwXnAwaq6v6r+Evg4cOkM6pAkMZsg\nOA34XyP3H+zbJEkzMPXB4iSvBl5aVf+ov//LwHlV9YZl610JXNnf/Qnga1Mt9OicCnxr1kUMwH7N\nl+3Yr+3YJxiuX3+zqhbWW+mYAXa8ngeBZ43cPx345vKVquoa4JppFTVJSfaNM1I/b+zXfNmO/dqO\nfYLZ92sWl4buAM5KckaS44DLgRtmUIckiRmcEVTVY0leD9wE7AA+VFX3TrsOSVJnFpeGqKrPAp+d\nxb6nZC4vaY3Bfs2X7div7dgnmHG/5uKTxZKk4TjXkCQ1ziDYpCQPJPlqkjuT7OvbTklyc5ID/e3J\nfXuS/E4/pcbdSZ432+qfkORDSQ4nuWekbcP9SLK7X/9Akt2z6MtILSv16R1J/nd/vO5M8vKRx36r\n79PXkrx0pH1LTYWS5FlJbk1yX5J7k7ypb5/b47VGn+b6eCU5PsmXktzV9+uqvv2MJLf3r/u1/Rtm\nSPLk/v7B/vHFkW2t2N+Jqip/NvEDPACcuqzt3wF7+uU9wDv75ZcD/xUIcD5w+6zrH6n5Z4HnAfds\nth/AKcD9/e3J/fLJW6xP7wDeusK65wB3AU8GzgC+Tvcmhh398pnAcf0658z4WO0EntcvPwX4877+\nuT1ea/Rpro9X/5qf2C8fC9zeH4NPAJf37e8H/km//BvA+/vly4Fr1+rvpOv1jGCyLgX29st7gVeM\ntP9+df4UOCnJzlkUuFxVfQH49rLmjfbjpcDNVfXtqvoOcDPwsuGrX9kqfVrNpcDHq+rRqvofwEG6\naVC23FQoVXWoqr7cL38fuI/uU/lze7zW6NNq5uJ49a/5D/q7x/Y/BVwIXNe3Lz9WS8fwOuCiJGH1\n/k6UQbB5Bfxxkv3pPgUN8MyqOgTdLzjwjL593qbV2Gg/5qV/r+8vkXxo6fIJc9qn/tLBc+n+p7kt\njteyPsGcH68kO5LcCRymC9uvAw9X1WP9KqM1/rD+/vHvAk9nSv0yCDbvgqp6Ht0sqq9L8rNrrJsV\n2ubx7Vqr9WMe+vc+4G8B5wKHgH/ft89dn5KcCHwSeHNVfW+tVVdo25J9W6FPc3+8qurxqjqXbvaE\n84CzV1qtv51pvwyCTaqqb/a3h4E/pDvQDy1d8ulvD/erjzWtxhay0X5s+f5V1UP9P8y/An6XJ06v\n56pPSY6l+4P50ar6VN8818drpT5tl+MFUFUPA7fRjRGclGTp81ujNf6w/v7xp9Fd3pxKvwyCTUjy\n15M8ZWkZuBi4h26qjKV3YOwGru+XbwB+pX8Xx/nAd5dO5beojfbjJuDiJCf3p/AX921bxrIxmVfS\nHS/o+nR5/66NM4CzgC+xBadC6a8ZfxC4r6reNfLQ3B6v1fo078cryUKSk/rlE4AX041/3Apc1q+2\n/FgtHcPLgM9VN1q8Wn8na1aj6vP8Q/fOhLv6n3uBt/ftTwduAQ70t6fUE+8geC/dNcKvArtm3YeR\nvnyM7tT7/9L97+OKzfQDeC3dQNZB4Ne2YJ8+0td8N90/rp0j67+979PXgJ8baX853btYvr50jGfc\nr5+huyxwN3Bn//PyeT5ea/Rpro8X8FPAV/r67wH+Vd9+Jt0f8oPAHwBP7tuP7+8f7B8/c73+TvLH\nTxZLUuO8NCRJjTMIJKlxBoEkNc4gkKTGGQSS1DiDQNtGksf7mSrvSfLppfdxT2C7v5rkP43bPqH9\n/djI/QeSnDrp/UhLDAJtJ49U1blV9ZN0n8p83awL2qRfBX5svZWkSTEItF19kZHJuZK8Lckd/SRm\nV420/5d+4sB7RyYPJMmvJfnzJJ8HLlhvZ/0nST/Z7+OOJBf07e/oJ027Lcn9Sd448px/meTP0n2H\nwMeSvDXJZcAu4KP92c0J/epvSPLldN+B8ZyjfnWkEQaBtp0kO4CL6KcYSHIx3Ufzz6ObxOz5I5ME\nvraqnk/3x/eNSZ7eT29wFV0AvIRuTvj1/Afg3VX1AuDvAx8Yeew5dFM/nwf86yTHJtnVr/dc4FX9\n/qmq64B9wGv6s5tH+m18q7pJDt8HvHWjr4m0lpl8eb00kBP6aX8Xgf10U/9CN5fOxXQf+Qc4kS4Y\nvkD3x/+Vffuz+va/AdxWVUcAklwL/O119v1i4Jxu6hwAnro0HxVwY1U9Cjya5DDwTLqpFa5f+kOf\n5NPrbH9pgrn9dMEhTYxBoO3kkao6N8nTgM/QjRH8Dt2cO/+2qv7z6MpJXkT3B/yFVfUXSW6jm/MF\nNj7V75P67Twy2tgHw6MjTY/T/btbaXrhtSxtY+n50sR4aUjbTlV9F3gj8NZ+iuObgNf2c96T5LQk\nz6Cb6vc7fQg8h26aYOi+GOVF/WWiY4FXj7HbPwZev3QnybnrrP8nwC+k+27bE4FLRh77Pt3XNkpT\n4f8stC1V1VeS3EX3/bAfSXI28MX+f+g/AP4h8EfArye5m25mxz/tn3soyTvoBpwPAV+m+07ctbwR\neG+/rWPoLjv9+hr13ZHkBroZbL9BNy7w3f7hDwPvT/II8MINdl3aMGcflWYkyYlV9YMkf40uOK6s\n/vt7pWnyjECanWuSnEM3LrHXENCseEYgSY1zsFiSGmcQSFLjDAJJapxBIEmNMwgkqXEGgSQ17v8B\nIPoEE55to/MAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1093a8208>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(df['seq'].apply(len), bins=500)\n",
    "plt.xlabel('Read length')\n",
    "plt.ylabel('Counts')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select reads that contain the primer in their sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "500    CGCACTACTTAGGTACTTCTaaaatATAATGTTACGTCCGAGACCT...\n",
      "501    GTGGGTTGCACGAGCTTTTAGACGCCAACGCATGAATaaaatAACG...\n",
      "502    ATCCTATGGGTGCCGAGTCAGCAAAAAATATCCGCCGCCCATATAA...\n",
      "503    GGTGTGCCCGGACTTGTCATGGACaaaatAGGTACTGGGAAGGCAG...\n",
      "504    AAGCAGGTCTGTGCGCCaaaatTTCGTTACCGCGTCAATTAGAAGC...\n",
      "Name: seq, dtype: object\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 1000 entries, 0 to 999\n",
      "Data columns (total 2 columns):\n",
      "name    1000 non-null object\n",
      "seq     1000 non-null object\n",
      "dtypes: object(2)\n",
      "memory usage: 23.4+ KB\n"
     ]
    }
   ],
   "source": [
    "boo = [] # Create empty list\n",
    "for i in df['seq']: # For each read in the dataframe\n",
    "    boo.append(forward in i and RC(reverse) in i) # Append a boolean indicating if both primer sequences (forward and reverse complement of reverse) are found in the read\n",
    "for_roi = df[boo].copy() # Subset the dataframe with only \"True\" reads / \n",
    "                         # .copy() avoids Panda's \"SettingwithCopyWarning\"\n",
    "\n",
    "# Perform the same selection for reverse complement reads\n",
    "boo = []\n",
    "for i in df['seq']:\n",
    "    boo.append(RC(forward) in i and reverse in i)\n",
    "rev_roi = df[boo].copy()\n",
    "\n",
    "rev_roi['seq'] = rev_roi['seq'].apply(RC) # Transform reverse reads into forward reads\n",
    "\n",
    "# Let's make sure the reverse reads are properly reversed\n",
    "print(rev_roi['seq'].head())\n",
    "\n",
    "roi = pd.concat([for_roi, rev_roi]) # Concatenate the 2 subsets\n",
    "roi.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Trim the reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aaaatAACACACGTTAGAACCGTATCGAGCATACAGCGGTGTTTCACCGGTGTTCACACCTCGCACCTGCCAAGACTACGGAGCCCCGTCGATCATAAACAGTGAGGCTGAGGCCGAGTGGCACCACGGTTGCGAGCCGGGCAGCCCGATCACATCGTTCCTTAAACACATAGGTACAATATCTCGCGTTACTGGTAAATATGTGTCAAATTAAGATCCTTAAGAGGCTGAGGCGTTTACTACTACGCGTTAAGGAATGTATGTGAACGAGACTGAATGATTTTCACTTTCGTAATCATGATAATACGTTGGCTGTGTCATCTGCGCTGGAGAATGTAGCAGCTTTGCTTTTTTGTGAGCAGAAGCTTCCACGAGGTGATTGGGAATGGCCTATGAATTA\n"
     ]
    }
   ],
   "source": [
    "roi['short'] = roi.apply(lambda x: shorten_read(read=x['seq'], primer=forward, n=max_len), axis=1) # Apply the trimming function to the column containing the read sequences and create a new column for the trimmed reads\n",
    "\n",
    "roi = roi[roi['short'].apply(len) >= min_len] # Select reads that are longer than the minimum length\n",
    "roi = roi[[\"name\", \"short\"]] # Keep only the read names and the short sequences\n",
    "\n",
    "# Make sure the trimming worked properly\n",
    "print(roi.iloc[0,1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot distribution of read length in the final dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYgAAAEKCAYAAAAIO8L1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAE0lJREFUeJzt3X+w5XV93/HnS5ZfGZRVWA3ZxV4o2xHGSVBXipppCZhU\nMQ1qJcGxEQ3pThITk1BNSDutodNOcdIRok21VFRwrELQhBVILBVW64wiu/xSi4aVwbAD464VUEY0\nA777x/dz9WT57N6zu/fcc+69z8fMmfP9fj6fc87ns9+753W+3+85n2+qCkmS9vS0aXdAkjSbDAhJ\nUpcBIUnqMiAkSV0GhCSpy4CQJHUZEJKkLgNCktRlQEiSutZMuwMH49hjj625ublpd0OSlpXt27d/\nq6rWLdRuWQfE3Nwc27Ztm3Y3JGlZSfKNcdp5iEmS1GVASJK6DAhJUpcBIUnqMiAkSV0GhCSpy4CQ\nJHUZEJKkLgNCktS1rH9JLUnL1dxFNxzU4++/5FWL1JO9cw9CktRlQEiSugwISVKXASFJ6jIgJEld\nBoQkqcuAkCR1GRCSpC4DQpLUZUBIkroMCElSlwEhSeoyICRJXQaEJKnLgJAkdRkQkqQuA0KS1GVA\nSJK6DAhJUpcBIUnqMiAkSV0GhCSpy4CQJHVNPCCSHJLkjiTXt/UTktya5N4kVyc5rJUf3tZ3tPq5\nSfdNkrR3S7EH8bvAPSPr7wQuraqNwMPABa38AuDhqjoJuLS1kyRNyUQDIskG4FXA+9t6gDOBa1uT\nK4FXt+Vz2jqt/qzWXpI0BZPeg7gM+APgh239GOCRqnqire8E1rfl9cADAK3+0dZekjQFEwuIJL8I\n7Kqq7aPFnaY1Rt3o825Osi3Jtt27dy9CTyVJPZPcg3gZ8EtJ7gc+xnBo6TJgbZI1rc0G4MG2vBM4\nHqDVHw18e88nrarLq2pTVW1at27dBLsvSavbxAKiqv6oqjZU1RxwHnBzVb0BuAV4XWt2PnBdW97S\n1mn1N1fVU/YgJElLYxq/g/hD4MIkOxjOMVzRyq8AjmnlFwIXTaFvkqRmzcJNDl5VbQW2tuX7gNM6\nbb4PnLsU/ZEkLcxfUkuSugwISVKXASFJ6jIgJEldBoQkqcuAkCR1GRCSpC4DQpLUZUBIkroMCElS\nlwEhSeoyICRJXQaEJKnLgJAkdRkQkqQuA0KS1GVASJK6DAhJUpcBIUnqMiAkSV0GhCSpy4CQJHUZ\nEJKkLgNCktRlQEiSugwISVKXASFJ6jIgJEldBoQkqcuAkCR1GRCSpC4DQpLUZUBIkroMCElSlwEh\nSeoyICRJXRMLiCRHJPlikruSfCXJxa38hCS3Jrk3ydVJDmvlh7f1Ha1+blJ9kyQtbJJ7ED8Azqyq\nnwFOBV6R5HTgncClVbUReBi4oLW/AHi4qk4CLm3tJElTMrGAqMFjbfXQdivgTODaVn4l8Oq2fE5b\np9WflSST6p8kad8meg4iySFJ7gR2ATcBXwceqaonWpOdwPq2vB54AKDVPwocM8n+SZL2bqIBUVVP\nVtWpwAbgNODkXrN239tbqD0LkmxOsi3Jtt27dy9eZyVJf8+SfIupqh4BtgKnA2uTrGlVG4AH2/JO\n4HiAVn808O3Oc11eVZuqatO6desm3XVJWrUm+S2mdUnWtuUjgZcD9wC3AK9rzc4HrmvLW9o6rf7m\nqnrKHoQkaWmsWbjJATsOuDLJIQxBdE1VXZ/k/wIfS/IfgTuAK1r7K4APJ9nBsOdw3gT7JklawMQC\noqruBl7QKb+P4XzEnuXfB86dVH8kSfvHX1JLkroMCElSlwEhSeoyICRJXQaEJKlrvwMiyTOT/PQk\nOiNJmh1jBUSSrUmekeRZwF3AB5O8a7JdkyRN07h7EEdX1XeA1wIfrKoXMfwyWpK0Qo0bEGuSHAf8\nMnD9BPsjSZoR4wbExcCngB1VdVuSE4F7J9ctSdK0jTvVxkNV9aMT01V1n+cgJGllG3cP4j1jlkmS\nVoh97kEkeQnwUmBdkgtHqp4BHDLJjkmSpmuhQ0yHAUe1dk8fKf8OP76mgyRpBdpnQFTVZ4DPJPlQ\nVX1jifokSZoB456kPjzJ5cDc6GOq6sxJdEqSNH3jBsSfA+8D3g88ObnuSJJmxbgB8URVvXeiPZEk\nzZRxv+b6ySS/leS4JM+av020Z5KkqRp3D+L8dv/2kbICTlzc7kiSZsVYAVFVJ0y6I5Kk2TJWQCR5\nY6+8qq5a3O5IkmbFuIeYXjyyfARwFnA7YEBI0go17iGm3xldT3I08OGJ9EiSNBMO9JrU3wM2LmZH\nJEmzZdxzEJ9k+NYSDJP0nQxcM6lOSZKmb9xzEP9lZPkJ4BtVtXMC/ZEkzYixDjG1Sfu+yjCj6zOB\nv5tkpyRJ0zdWQCT5ZeCLwLkM16W+NYnTfUvSCjbuIaZ/C7y4qnYBJFkH/G/g2kl1TJI0XeN+i+lp\n8+HQ/L/9eKwkaRkadw/ir5N8CvhoW/8V4MbJdEmSNAsWuib1ScBzqurtSV4L/CwQ4PPAR5agf5Kk\nKVnoMNFlwHcBquoTVXVhVf0+w97DZZPunCRpehYKiLmqunvPwqraxnD5UUnSCrVQQByxj7ojF7Mj\nkqTZslBA3JbkX+1ZmOQCYPu+Hpjk+CS3JLknyVeS/G4rf1aSm5Lc2+6f2cqT5N1JdiS5O8kLD3RQ\nkqSDt9C3mH4P+Iskb+DHgbAJOAx4zQKPfQL411V1e5KnA9uT3AS8Cfh0VV2S5CLgIuAPgVcyTAC4\nEfjHwHvbvSRpCvYZEFX1TeClSX4OeH4rvqGqbl7oiavqIeChtvzdJPcA64FzgDNasyuBrQwBcQ5w\nVVUV8IUka5Mc155HkrTExr0exC3ALQf6IknmgBcAtzJ8bXY+OB5K8uzWbD3wwMjDdrayvxcQSTYD\nmwGe+9znHmiXJEkLmPivoZMcBXwc+L2q+s6+mnbK6ikFVZdX1aaq2rRu3brF6qYkaQ8TDYgkhzKE\nw0eq6hOt+JtJjmv1xwHzU3jsBI4fefgG4MFJ9k+StHcTC4gkAa4A7qmqd41UbQHOb8vnA9eNlL+x\nfZvpdOBRzz9I0vSMOxfTgXgZ8KvAl5Lc2cr+DXAJcE37quzfMkwhDsOvs88GdjBc0vTNE+ybJGkB\nEwuIqvoc/fMKAGd12hfwlkn1R5K0f5yyW5LUZUBIkroMCElSlwEhSeoyICRJXQaEJKnLgJAkdRkQ\nkqQuA0KS1GVASJK6DAhJUpcBIUnqMiAkSV0GhCSpy4CQJHUZEJKkLgNCktRlQEiSugwISVKXASFJ\n6jIgJEldBoQkqcuAkCR1GRCSpC4DQpLUZUBIkroMCElSlwEhSeoyICRJXQaEJKnLgJAkdRkQkqQu\nA0KS1GVASJK6DAhJUtfEAiLJB5LsSvLlkbJnJbkpyb3t/pmtPEnenWRHkruTvHBS/ZIkjWeSexAf\nAl6xR9lFwKeraiPw6bYO8EpgY7ttBt47wX5JksYwsYCoqs8C396j+BzgyrZ8JfDqkfKravAFYG2S\n4ybVN0nSwpb6HMRzquohgHb/7Fa+HnhgpN3OViZJmpJZOUmdTll1Gyabk2xLsm337t0T7pYkrV5L\nHRDfnD901O53tfKdwPEj7TYAD/aeoKour6pNVbVp3bp1E+2sJK1mSx0QW4Dz2/L5wHUj5W9s32Y6\nHXh0/lCUJGk61kzqiZN8FDgDODbJTuAdwCXANUkuAP4WOLc1vxE4G9gBfA9486T6JUkaz8QCoqpe\nv5eqszptC3jLpPoiSdp/s3KSWpI0YwwISVKXASFJ6jIgJEldBoQkqcuAkCR1GRCSpC4DQpLUZUBI\nkroMCElSlwEhSeoyICRJXQaEJKnLgJAkdRkQkqQuA0KS1GVASJK6DAhJUpcBIUnqMiAkSV0GhCSp\ny4CQJHUZEJKkLgNCktRlQEiSugwISVKXASFJ6jIgJEldBoQkqWvNtDsg6anmLrrhgB97/yWvWsSe\naDVbtQFxMP8Bwf+E2reD/ftajpbr/ynDeO9WbUBodfA/v3TgDAhJM8Ewnz0GhMayXA8faHVYjYf0\nloIBcYCW4x+kb9L7ZzluY1i+/dbsMSBWkWm+cfimJS0/MxUQSV4B/ClwCPD+qrpkyl2SpL1a6R98\nZuaHckkOAf4MeCVwCvD6JKdMt1eStHrNTEAApwE7quq+qvo74GPAOVPukyStWrMUEOuBB0bWd7Yy\nSdIUzNI5iHTK6imNks3A5rb6WJKvHeDrHQt86wAfu1w55tXBMa8CeedBjfkfjNNolgJiJ3D8yPoG\n4ME9G1XV5cDlB/tiSbZV1aaDfZ7lxDGvDo55dViKMc/SIabbgI1JTkhyGHAesGXKfZKkVWtm9iCq\n6okkvw18iuFrrh+oqq9MuVuStGrNTEAAVNWNwI1L9HIHfZhqGXLMq4NjXh0mPuZUPeU8sCRJM3UO\nQpI0Q1ZsQCQ5IskXk9yV5CtJLm7lJyS5Ncm9Sa5uJ8RJcnhb39Hq56bZ//21j/F+JMnXknw5yQeS\nHNrKk+Tdbbx3J3nhdEew//Y25pH69yR5bGR9WW9j2Od2TpL/lORvktyT5K0j5StyOyc5K8ntSe5M\n8rkkJ7XyZb+d5yU5JMkdSa5v60v7/lVVK/LG8LuKo9ryocCtwOnANcB5rfx9wG+25d8C3teWzwOu\nnvYYFmm8Z7e6AB8dGe/ZwF+18tOBW6c9hsUac1vfBHwYeGyk/bLexgts5zcDVwFPa3XPXunbGfgb\n4OSRbfuhlbKdR8Z+IfA/gevb+pK+f63YPYgazH96PLTdCjgTuLaVXwm8ui2f09Zp9Wcl6f14bybt\nbbxVdWOrK+CLDL8vgWG8V7WqLwBrkxy39D0/cHsbc5vX60+AP9jjIct6G8M+/65/E/gPVfXD1m5X\na7Nit3O7PaOVH82Pfze17LczQJINwKuA97f1sMTvXys2IOBHu2d3AruAm4CvA49U1ROtyeh0Hj+a\n6qPVPwocs7Q9Pjh7jreqbh2pOxT4VeCvW9GKmNpkL2P+bWBLVT20R/Nlv41hr2P+h8CvJNmW5K+S\nbGzNV/J2/nXgxiQ7Gf6252d/XhHbGbiM4UPOD9v6MSzx+9eKDoiqerKqTmX41HwacHKvWbsfa6qP\nWbbneJM8f6T6vwGfrar/09aX/XihO+Z/ApwLvKfTfKWO+fnA4cD3a/hl7f8APtCar+Qx/z5wdlVt\nAD4IvKs1X/ZjTvKLwK6q2j5a3Gk60fevFR0Q86rqEWArw3HLtUnmf/8xOp3Hj6b6aPVHA99e2p4u\njpHxvgIgyTuAdQzHM+eNNbXJcjEy5p8DTgJ2JLkf+IkkO1qzFbON4SnbeSfw8Vb1F8BPt+WVup1f\nCfzMyF7y1cBL2/JK2M4vA36p/Q1/jOHQ0mUs8fvXig2IJOuSrG3LRwIvB+4BbgFe15qdD1zXlre0\ndVr9ze24/bKwl/F+NcmvA/8MeP388elmC/DG9i2X04FHO4dkZtpexry9qn6yquaqag74XlWd1B6y\nrLcx7H07A3/J8CYC8E8ZTuDCyt3O9wBHJ/lHrdnPtzJYAdu5qv6oqja0v+HzGMbwBpb6/WsxznTP\n4o3hE9QdwN3Al4F/38pPZDhZuwP4c+DwVn5EW9/R6k+c9hgWabxPMJx7ubPd5svDcIGmrwNfAjZN\newyLNeY92ox+i2lZb+MFtvNa4Ia2LT/P8Ol6RW9n4DVtTHcx7FWcuFK28x7jP4Mff4tpSd+//CW1\nJKlrxR5ikiQdHANCktRlQEiSugwISVKXASFJ6jIgtOIlebLN+PnlJJ+c/079Ijzvm5L813HLF+n1\nfmpk/f4kxy7260jzDAitBo9X1alV9XyGX5e+ZdodOkBvAn5qoUbSYjEgtNp8npHJ6pK8Pclt7VoJ\nF4+U/2WS7e36A5tHyt+c4ZoLn2GYDmGf2q+AP95e47YkL2vlf5zh+hxbk9yXdv2GVvfvknw1yU1J\nPprkbUlexzCF+Ufa3tCRrfnvZLgmwpeSPO+g/3WkEQaEVo02DfhZDNMSkOQXgI0MEzmeCryoTfYH\n8GtV9SKGN+W3JjmmTZN9MUMw/Dxwyhgv+6fApVX1YuBf0KZubp7HMA3KacA7khyaZFNr9wLgte31\nqaprgW3AG9re0OPtOb5VVS8E3gu8bX//TaR9WbNwE2nZO7JNFT0HbGeY+h3gF9rtjrZ+FENgfJYh\nFF7Tyo9v5T8JbK2q3QBJrgbm5wLam5cDp4xMzf+MJE9vyzdU1Q+AHyTZBTwH+FnguvkASPLJBZ7/\nE+1+O0OgSIvGgNBq8HhVnZrkaOB6hnMQ72aYp+g/V9V/H22c5AyGN/aXVNX3kmxlmOsG9n8K5ae1\n53l8tLAFxg9Gip5k+P+4vxd5mX+O+cdLi8ZDTFo1qupR4K3A29oFlD4F/FqSowCSrE/ybIapkh9u\n4fA8hmniYbjU5RntcNOhDNedWMj/YriAEe01Tl2g/eeAf57hOsxHMVxRbN53gaf3HyYtPj9xaFWp\nqjuS3MVwXd8PJzkZ+Hz7RP8Y8C8Zrrr3G0nuBr4GfKE99qEkf8xwovsh4HbgkAVe8q3An7XnWsNw\n+Oo39tG/25JsYZih9BsM5x0ebdUfAt6X5HHgJfs5dGm/OZurNGOSHFVVjyX5CYZA2VxVt0+7X1p9\n3IOQZs/lSU5hOO9xpeGgaXEPQpLU5UlqSVKXASFJ6jIgJEldBoQkqcuAkCR1GRCSpK7/D6GFx9VY\ne5IwAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10f6d3d30>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(roi['short'].apply(len), bins=20)\n",
    "plt.xlabel('Read length')\n",
    "plt.ylabel('Counts')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create output fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "output = roi.apply(lambda x: f\">{x[0]}\\n{x[1]}\\n\", axis=1) # Put read name and sequence in fasta format\n",
    "fa = \"\"\n",
    "for i in output:\n",
    "    fa = fa + i # Append all reads in fasta format into a single string\n",
    "\n",
    "with open(outfile, \"w\") as f:\n",
    "    f.write(fa)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
